perceval/backends/kitsune.py

```diff
@@ -0,0 +1,383 @@
+    """
+    version = '0.1.0'
+
+    def __init__(self, url=None, cache=None, origin=None):
+        if not url:
+            url = KITSUNE_URL
+        origin = origin if origin else url
+
+        super().__init__(origin, cache=cache)
+        self.url = url
+        self.client = KitsuneClient(url)
+        self.cache_offset = 0
+
+    @kitsune_metadata
+    @metadata
+    def fetch(self, offset=DEFAULT_OFFSET):
+        """Fetch questions from the Kitsune url.
+
+        :offset: obtain questions after offset
+        :returns: a generator of questions
+        """
+        logger.info("Looking for questions at url '%s' using offset %s",
+                    self.url, str(offset))
+
+        self._purge_cache_queue()
+        # Add to the cache the offset so it can be used to recover from cache
+        self._push_cache_queue(offset)
+        self._flush_cache_queue()
+
+        nquestions = 0  # number of questions processed
+        tquestions = 0  # number of questions from API data
+        drop_questions = 0 # drop questions from page before the offset
+
+        # Always get complete pages so the first item is always
+        # the first one in the page
+        page = int(offset/KitsuneClient.ITEMS_PER_PAGE)
+        page_offset = page * KitsuneClient.ITEMS_PER_PAGE
+        drop_questions = offset - page_offset
+        current_offset = offset
+
+        for raw_questions in self.client.get_questions(offset):
+            self._push_cache_queue(raw_questions)
+            try:
+                questions_data = json.loads(raw_questions)
+                tquestions = questions_data['count']
+                questions = questions_data['results']
+            except (ValueError, KeyError) as ex:
+                logger.error(ex)
+                cause = ("Bad JSON format for mozilla_questions: %s" % (raw_questions))
+                raise ParseError(cause=cause)
+
+            for question in questions:
+                if drop_questions > 0:
+                    # Remove extra questions due to page base retrieval
+                    drop_questions -= 1
+                    continue
+                question['offset'] = current_offset
+                current_offset += 1
+                question['answers_data'] = []
+                for raw_answers in self.client.get_question_answers(question['id']):
+                    self._push_cache_queue(raw_answers)
+                    answers = json.loads(raw_answers)['results']
+                    question['answers_data'] += answers
+                yield question
+                nquestions += 1
+                self._push_cache_queue('{}') # Mark with empty dict end of question
+
+            logger.debug("Questions: %i/%i", nquestions, tquestions)
+
+            self._flush_cache_queue()
+
+        logger.info("Total number of questions: %i (%i total)", nquestions, tquestions)
+
+    @kitsune_metadata
+    @metadata
+    def fetch_from_cache(self):
```

Comment:
This method is too long. I think you can split it in several private methods to make it more readable.